loki/transformations/inline.py

```diff
@@ -339,6 +360,89 @@ def inline_elemental_functions(routine):
+        nesting that can be replaced.
+    """
+    def apply_to_init_arg(name, arg, expr, mapper):
+        # Helper utility to apply the mapper only to expression arguments and
+        # retain the scope while rebuilding the node
+        if isinstance(arg, (tuple, Expression)):
+            return mapper(arg)
+        if name == 'scope':
+            return expr.scope
+        return arg
+
+    for _ in range(max_iterations):
+        # We update the expression map by applying it to the children of each replacement
+        # node, thus making sure node replacements are also applied to nested attributes,
+        # e.g. call arguments or array subscripts etc.
+        mapper = InlineSubstitutionMapper(expr_map) # SubstituteExpressionsMapper(expr_map)
+        prev_map, expr_map = expr_map, {
+            expr: type(replacement)(**{
+                name: apply_to_init_arg(name, arg, expr, mapper)
+                for name, arg in zip(replacement.init_arg_names, replacement.__getinitargs__())
+            })
+            for expr, replacement in expr_map.items()
+        }
+
+        # Check for early termination opportunities
+        if prev_map == expr_map:
+            break
+
+    return expr_map
+
+
+def inline_statement_functions(routine):
+    """
+    Replaces `InlineCall` expression to statement functions with the
+    called statement functions rhs.
+    """
+    # Keep track of removed symbols
+    removed_functions = set()
+
+    stmt_func_decls = FindNodes(StatementFunction).visit(routine.spec)
+    exprmap = {}
+    for call in FindInlineCalls().visit(routine.body):
+        if call.procedure_type is BasicType.DEFERRED:
+            continue
+        if call.procedure_type.is_function and isinstance(call.routine, StatementFunction):
+            exprmap[call] = InlineSubstitutionMapper()(call, scope=routine)
+            removed_functions.add(call.routine)
+    # Apply the map to itself to handle nested statement function calls
+    exprmap = recursive_inline_expression_map_update(exprmap, max_iterations=10)
+    # Apply expression-level substitution to routine
+    routine.body = SubstituteExpressions(exprmap).visit(routine.body)
+
+    # collect all the arguments of the statement functions
+    stmt_arguments = set()
+    stmt_func_map = {}
+    for stmt_func_decl in stmt_func_decls:
+        for arg in stmt_func_decl.arguments:
+            stmt_arguments.add(arg)
+        stmt_func_map[stmt_func_decl] = None
+    # remove the statement function declarations
+    routine.spec = Transformer(stmt_func_map).visit(routine.spec)
+
+    # remove the declarations of the statement function arguments
+    decls = FindNodes(VariableDeclaration).visit(routine.spec)
+    decls = tuple(d for d in decls if any(isinstance(s, sym.ProcedureSymbol) or s in stmt_arguments for s in d.symbols))
+    decl_map = {}
+    for decl in decls:
+        decl_map[decl] = None
+    routine.spec = Transformer(decl_map).visit(routine.spec)
```

Comment:
This boils down to removing a bunch of variable declarations and everything in `stmt_func_decls`, so that should be possible in a single `Transformer` pass. Maybe something along these lines works?
```python
vars_to_remove = {stmt_func.variable.name.lower() for stmt_func in stmt_func_decls}
vars_to_remove += {arg.name.lower() for stmt_func in stmt_func_decls for arg in stmt_func.arguments}
spec_map = {stmt_func: None for stmt_func in stmt_func_decls}
for decl in routine.declarations:
    if any(var in vars_to_remove for var in decl.symbols):
        symbols = tuple(var for var in decl.symbols if var not in vars_to_remove)
        if symbols:
            decl._update(symbols=symbols)
        else:
            spec_map[decl] = None
routine.spec = Transformer(spec_map).visit(routine.spec)
```